### DISCUSSION TITLE
Compiling recursive methods

### DISCUSSION THREAD

**Discussion by sebastienros** [dadhi/FastExpressionCompiler#293] [Q&A]
In the default implementation it's not possible to "compile" recursive methods. I think it fails at runtime last time I tried.
Do you think FEC would support that?

**Comment by dadhi** [comment]
I am not sure, but don't think anything prevents it. Did you try it with FEC?
Or do you mean the tail-recursive methods?

**Reply by jogibear9988** [reply]
seems not to work....
I've added a test #294

**Reply by jogibear9988** [reply]
I used a Expression from here: https://chriscavanagh.wordpress.com/2012/06/18/recursive-methods-in-expression-trees/

IL of system Expression

	0    ldc.i4.1
	1    newarr Object
	6    dup
	7    ldc.i4.0
	8    newobj StrongBox`1..ctor
	13   stelem.ref
	14   stloc.0
	15   ldarg.0
	16   ldfld Closure.Constants
	21   ldc.i4.0
	22   ldelem.ref
	23   castclass MethodInfo
	28   ldtoken Func`2
	33   call Type.GetTypeFromHandle
	38   ldnull
	39   ldloc.0
	40   newobj Closure..ctor
	45   callvirt MethodInfo.CreateDelegate
	50   castclass Func`2
	55   stloc.1
	56   ldloc.0
	57   ldc.i4.0
	58   ldelem.ref
	59   castclass StrongBox`1
	64   ldloc.1
	65   stfld StrongBox`1.Value
	70   ldloc.0
	71   ldc.i4.0
	72   ldelem.ref
	73   castclass StrongBox`1
	78   ldfld StrongBox`1.Value
	83   ldarg.1
	84   callvirt Func`2.Invoke
	89   ret

IL of FEC

	0    ldarg.0
	1    ldfld ArrayClosure.ConstantsAndNestedLambdas
	6    stloc.0
	7    ldloc.0
	8    ldc.i4.0
	9    ldelem.ref
	10   stloc.1
	11   ldloc.1
	12   ldc.i4.1
	13   newarr Object
	18   dup
	19   ldc.i4.0
	20   ldloc.2
	21   stelem.ref
	22   newobj ArrayClosureWithNonPassedParams..ctor
	27   call CurryClosureFuncs.Curry
	32   stloc.2
	33   ldloc.2
	34   ldarg.1
	35   call Func`2.Invoke
	40   ret

**Reply by sebastienros** [reply]
@jogibear9988 this is the only sample on Internet that I could find too ;) I used it to implement recursion in my project.

**Reply by sebastienros** [reply]
Looking at this SO answer, the program might not be valid. The recursive call of the lambda might require a closure class.

When I write this in a console app, ILSpy shows me that a class is created to hold the lambda:

```
            Func<uint, uint> factorial = null;
            factorial = n => (n <= 1) ? 1 : n * factorial(n - 1);
```

```
.method private hidebysig static 
	void Main (
		string[] args
	) cil managed 
{
	// Method begins at RVA 0x2050
	// Code size 51 (0x33)
	.maxstack 3
	.entrypoint
	.locals init (
		[0] class ConsoleApp4.Program/'<>c__DisplayClass0_0' 'CS$<>8__locals0'
	)

	// {
	IL_0000: newobj instance void ConsoleApp4.Program/'<>c__DisplayClass0_0'::.ctor()
	IL_0005: stloc.0
	// (no C# code)
	IL_0006: nop
	// Func<uint, uint> factorial = null;
	IL_0007: ldloc.0
	IL_0008: ldnull
	IL_0009: stfld class [System.Runtime]System.Func`2<uint32, uint32> ConsoleApp4.Program/'<>c__DisplayClass0_0'::factorial
	// factorial = (uint n) => (n <= 1) ? 1u : (n * factorial(n - 1));
	IL_000e: ldloc.0
	IL_000f: ldloc.0
	IL_0010: ldftn instance uint32 ConsoleApp4.Program/'<>c__DisplayClass0_0'::'<Main>b__0'(uint32)
	IL_0016: newobj instance void class [System.Runtime]System.Func`2<uint32, uint32>::.ctor(object, native int)
	IL_001b: stfld class [System.Runtime]System.Func`2<uint32, uint32> ConsoleApp4.Program/'<>c__DisplayClass0_0'::factorial
	// Console.WriteLine(factorial(3u));
	IL_0020: ldloc.0
	IL_0021: ldfld class [System.Runtime]System.Func`2<uint32, uint32> ConsoleApp4.Program/'<>c__DisplayClass0_0'::factorial
	IL_0026: ldc.i4.3
	IL_0027: callvirt instance !1 class [System.Runtime]System.Func`2<uint32, uint32>::Invoke(!0)
	IL_002c: call void [System.Console]System.Console::WriteLine(uint32)
	// }
	IL_0031: nop
	IL_0032: ret
} // end of method Program::Main

```

**Reply by jogibear9988** [reply]
But why does that mean the program is invalid? It works, you could compile with Microsofts Expression compiler.
If FEC could not compile it, it should throw an exception.

**Reply by sebastienros** [reply]
Fair, but based on the answer I wonder if the user code expression tree creation should be responsible for handling the closure class creation and usage. Maybe the standard compiler detects the closure and creates the plumbing, while FEC doesn't.

**Reply by jogibear9988** [reply]
Think it would be okay. It only should throw that you know why it wont work.
Don't know if recursion should be in scope of Expression Trees

**Reply by dadhi** [reply]
@jogibear9988 @sebastienros
I found the cause of the issue. FEC closure variables are immutable and cannot be set outside of the nested lambda after lambda is compiled (emitted and stored into the parent closure). You may hack it by using the array of one element as a variable.. check the code from 
here https://github.com/dadhi/FastExpressionCompiler/blob/95d1d39781d82bfdda55d634f298ea21cfb9224e/test/FastExpressionCompiler.IssueTests/Issue293_Recursive_Methods.cs#L51

Now I need to find a simple way to detect this situation and return null from the Compiler instead of throwing.

**Reply by sebastienros** [reply]
I also fixed it by creating a closure class that holds the Func in a field, and pass an instance of this class, like what the compiler is doing as explained in the SO answer by Eric Lippert. I can even reuse the same instance from multiple compiled lambdas.

**Reply by dadhi** [reply]
Good, I have opened an issue #295 to make it possible in FEC. I just need to create the closure items array and change the variable assignement to array element assignment and usage (for the variables passed to closure).
